pg_rewind.1
'\" t
.\"     Title: pg_rewind
.\"    Author: The PostgreSQL Global Development Group
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 2016
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PG_REWIND" "1" "2016" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pg_rewind \- synchronize a PostgreSQL data directory with another data directory that was forked from the first one
.SH "SYNOPSIS"
.HP \w'\fBpg_rewind\fR\ 'u
\fBpg_rewind\fR [\fIoption\fR...] {\fB\-D\ \fR | \fB\-\-target\-pgdata\fR}\fI directory\fR {\fB\-\-source\-pgdata=\fR\fB\fIdirectory\fR\fR | \fB\-\-source\-server=\fR\fB\fIconnstr\fR\fR}
.SH "DESCRIPTION"
.PP
pg_rewind
is a tool for synchronizing a PostgreSQL cluster with another copy of the same cluster, after the clusters\*(Aq timelines have diverged\&. A typical scenario is to bring an old master server back online after failover, as a standby that follows the new master\&.
.PP
The result is equivalent to replacing the target data directory with the source one\&. All files are copied, including configuration files\&. The advantage of
pg_rewind
over taking a new base backup, or tools like
rsync, is that
pg_rewind
does not require reading through all unchanged files in the cluster\&. That makes it a lot faster when the database is large and only a small portion of it differs between the clusters\&.
.PP
pg_rewind
examines the timeline histories of the source and target clusters to determine the point where they diverged, and expects to find WAL in the target cluster\*(Aqs
pg_xlog
directory reaching all the way back to the point of divergence\&. In the typical failover scenario where the target cluster was shut down soon after the divergence, that is not a problem, but if the target cluster had run for a long time after the divergence, the old WAL files might not be present anymore\&. In that case, they can be manually copied from the WAL archive to the
pg_xlog
directory\&. Fetching missing files from a WAL archive automatically is currently not supported\&.
.PP
When the target server is started up for the first time after running
pg_rewind, it will go into recovery mode and replay all WAL generated in the source server after the point of divergence\&. If some of the WAL was no longer available in the source server when
pg_rewind
was run, and therefore could not be copied by
pg_rewind
session, it needs to be made available when the target server is started up\&. That can be done by creating a
recovery\&.conf
file in the target data directory with a suitable
\fIrestore_command\fR\&.
.SH "OPTIONS"
.PP
pg_rewind
accepts the following command\-line arguments:
.PP
\fB\-D \fR\fB\fIdirectory\fR\fR
.br
\fB\-\-target\-pgdata=\fR\fB\fIdirectory\fR\fR
.RS 4
This option specifies the target data directory that is synchronized with the source\&. The target server must shut down cleanly before running
pg_rewind
.RE
.PP
\fB\-\-source\-pgdata=\fR\fB\fIdirectory\fR\fR
.RS 4
Specifies path to the data directory of the source server, to synchronize the target with\&. When
\fB\-\-source\-pgdata\fR
is used, the source server must be cleanly shut down\&.
.RE
.PP
\fB\-\-source\-server=\fR\fB\fIconnstr\fR\fR
.RS 4
Specifies a libpq connection string to connect to the source
PostgreSQL
server to synchronize the target with\&. The server must be up and running, and must not be in recovery mode\&.
.RE
.PP
\fB\-n\fR
.br
\fB\-\-dry\-run\fR
.RS 4
Do everything except actually modifying the target directory\&.
.RE
.PP
\fB\-v\fR
.br
\fB\-\-verbose\fR
.RS 4
Enables verbose reporting\&. Turning this on will deliver an approximate progress report while copying data over from the source cluster\&.
.RE
.PP
\fB\-V\fR
.br
\fB\-\-version\fR
.RS 4
Display version information, then exit
.RE
.PP
\fB\-?\fR
.br
\fB\-\-help\fR
.RS 4
Show help, then exit
.RE
.SH "ENVIRONMENT"
.PP
When
\fB\-\-source\-server\fR
option is used,
pg_rewind
also uses the environment variables supported by libpq.
.SH "NOTES"
.PP
pg_rewind
requires that the
\fIwal_log_hints\fR
option is enabled in
postgresql\&.conf, or that data checksums were enabled when the cluster was initialized with
initdb\&.
\fIfull_page_writes\fR
must also be enabled\&.
.SS "How it works"
.PP
The basic idea is to copy everything from the new cluster to the old cluster, except for the blocks that we know to be the same\&.
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Scan the WAL log of the old cluster, starting from the last checkpoint before the point where the new cluster\*(Aqs timeline history forked off from the old cluster\&. For each WAL record, make a note of the data blocks that were touched\&. This yields a list of all the data blocks that were changed in the old cluster, after the new cluster forked off\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Copy all those changed blocks from the new cluster to the old cluster\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  3." 4.2
.\}
Copy all other files like clog, conf files etc\&. from the new cluster to old cluster\&. Everything except the relation files\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 4.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  4." 4.2
.\}
Apply the WAL from the new cluster, starting from the checkpoint created at failover\&. (Strictly speaking,
pg_rewind
doesn\*(Aqt apply the WAL, it just creates a backup label file indicating that when
PostgreSQL
is started, it will start replay from that checkpoint and apply all the required WAL\&.)
.RE


pg_rewind.c
/*-------------------------------------------------------------------------
 *
 * pg_rewind.c
 *	  Synchronizes an old master server to a new timeline
 *
 * Portions Copyright (c) 1996-2014, PostgreSQL Global Development Group
 * Portions Copyright (c) 2013-2014 VMware, Inc. All Rights Reserved.
 *
 *-------------------------------------------------------------------------
 */
#include "postgres_fe.h"

#include "pg_rewind.h"
#include "fetch.h"
#include "filemap.h"

#include "access/timeline.h"
#include "access/xlog_internal.h"
#include "catalog/catversion.h"
#include "catalog/pg_control.h"
#include "storage/bufpage.h"

#include "getopt_long.h"

#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>
#include <unistd.h>

static void usage(const char *progname);

static void createBackupLabel(XLogRecPtr startpoint, TimeLineID starttli,
				  XLogRecPtr checkpointloc);

static void digestControlFile(ControlFileData *ControlFile, char *source, size_t size);
static void updateControlFile(ControlFileData *ControlFile,
							  char *datadir);
static void sanityChecks(void);
static void findCommonAncestorTimeline(XLogRecPtr *recptr, TimeLineID *tli);

static ControlFileData ControlFile_target;
static ControlFileData ControlFile_source;

const char *progname;

char *datadir_target = NULL;
char *datadir_source = NULL;
char *connstr_source = NULL;

int verbose;
int dry_run;

static void
usage(const char *progname)
{
	printf("%s resynchronizes a cluster with another copy of the cluster.\n\n", progname);
	printf("Usage:\n  %s [OPTION]...\n\n", progname);
	printf("Options:\n");
	printf("  -D, --target-pgdata=DIRECTORY\n");
	printf("                 existing data directory to modify\n");
	printf("  --source-pgdata=DIRECTORY\n");
	printf("                 source data directory to sync with\n");
	printf("  --source-server=CONNSTR\n");
	printf("                 source server to sync with\n");
	printf("  -v, --verbose  write a lot of progress messages\n");
	printf("  -n, --dry-run  stop before modifying anything\n");
	printf("  -V, --version  output version information, then exit\n");
	printf("  -?, --help     show this help, then exit\n");
	printf("\n");
	printf("Report bugs to https://github.com/vmware/pg_rewind.\n");
}


int
main(int argc, char **argv)
{
	static struct option long_options[] = {
		{"help", no_argument, NULL, '?'},
		{"target-pgdata", required_argument, NULL, 'D'},
		{"source-pgdata", required_argument, NULL, 1},
		{"source-server", required_argument, NULL, 2},
		{"version", no_argument, NULL, 'V'},
		{"dry-run", no_argument, NULL, 'n'},
		{"verbose", no_argument, NULL, 'v'},
		{NULL, 0, NULL, 0}
	};
	int			option_index;
	int			c;
	XLogRecPtr	divergerec;
	TimeLineID	lastcommontli;
	XLogRecPtr	chkptrec;
	TimeLineID	chkpttli;
	XLogRecPtr	chkptredo;
	size_t		size;
	char	   *buffer;
	bool		rewind_needed;
	ControlFileData	ControlFile;

	set_pglocale_pgservice(argv[0], PG_TEXTDOMAIN("pg_rewind"));
	progname = get_progname(argv[0]);

	/* Set default parameter values */
	verbose = 0;
	dry_run = 0;

	/* Process command-line arguments */
	if (argc > 1)
	{
		if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-?") == 0)
		{
			usage(progname);
			exit(0);
		}
		if (strcmp(argv[1], "--version") == 0 || strcmp(argv[1], "-V") == 0)
		{
			puts("pg_rewind " PG_REWIND_VERSION);
			exit(0);
		}
	}

	while ((c = getopt_long(argc, argv, "D:vn", long_options, &option_index)) != -1)
	{
		switch (c)
		{
			case '?':
				fprintf(stderr, _("Try \"%s --help\" for more information.\n"), progname);
				exit(1);
			case ':':
				exit(1);
			case 'v':
				verbose = 1;
				break;
			case 'n':
				dry_run = 1;
				break;

			case 'D':	/* -D or --target-pgdata */
				datadir_target = pg_strdup(optarg);
				break;

			case 1:		/* --source-pgdata */
				datadir_source = pg_strdup(optarg);
				break;
			case 2:		/* --source-server */
				connstr_source = pg_strdup(optarg);
				break;
		}
	}

	/* No source given? Show usage */
	if (datadir_source == NULL && connstr_source == NULL)
	{
		fprintf(stderr, "%s: no source specified\n", progname);
		fprintf(stderr, _("Try \"%s --help\" for more information.\n"), progname);
		exit(1);
	}

	if (datadir_target == NULL)
	{
		fprintf(stderr, "%s: no target data directory specified\n", progname);
		fprintf(stderr, _("Try \"%s --help\" for more information.\n"), progname);
		exit(1);
	}

	if (argc != optind)
	{
		fprintf(stderr, "%s: invalid arguments\n", progname);
		fprintf(stderr, _("Try \"%s --help\" for more information.\n"), progname);
		exit(1);
	}

	/*
	 * Don't allow pg_rewind to be run as root, to avoid overwriting the
	 * ownership of files in the data directory. We need only check for root
	 * -- any other user won't have sufficient permissions to modify files in
	 * the data directory.
	 */
#ifndef WIN32
	if (geteuid() == 0)
	{
		fprintf(stderr, "cannot be executed by \"root\"\n"
				"You must run %s as the PostgreSQL superuser.\n",
				progname);
		exit(1);
	}
#endif

	/*
	 * Connect to remote server
	 */
	if (connstr_source)
		libpqConnect(connstr_source);

	/*
	 * Ok, we have all the options and we're ready to start. Read in all the
	 * information we need from both clusters.
	 */
	buffer = slurpFile(datadir_target, "global/pg_control", &size);
	digestControlFile(&ControlFile_target, buffer, size);
	pg_free(buffer);

	buffer = fetchFile("global/pg_control", &size);
	digestControlFile(&ControlFile_source, buffer, size);
	pg_free(buffer);

	sanityChecks();

	/*
	 * If both clusters are already on the same timeline, there's nothing
	 * to do.
	 */
	if (ControlFile_target.checkPointCopy.ThisTimeLineID == ControlFile_source.checkPointCopy.ThisTimeLineID)
	{
		fprintf(stderr, "source and target cluster are both on the same timeline.\n");
		printf("No rewind required.\n");
		exit(0);
	}

	findCommonAncestorTimeline(&divergerec, &lastcommontli);
	printf("The servers diverged at WAL position %X/%X on timeline %u.\n",
		   (uint32) (divergerec >> 32), (uint32) divergerec, lastcommontli);

	/*
	 * Check for the possibility that the target is in fact a direct ancestor
	 * of the source. In that case, there is no divergent history in the
	 * target that needs rewinding.
	 */
	if (ControlFile_target.checkPoint >= divergerec)
	{
		rewind_needed = true;
	}
	else
	{
		XLogRecPtr chkptendrec;

		/* Read the checkpoint record on the target to see where it ends. */
		chkptendrec = readOneRecord(datadir_target,
									ControlFile_target.checkPoint,
									ControlFile_target.checkPointCopy.ThisTimeLineID);

		/*
		 * If the histories diverged exactly at the end of the shutdown
		 * checkpoint record on the target, there are no WAL records in the
		 * target that don't belong in the source's history, and no rewind is
		 * needed.
		 */
		if (chkptendrec == divergerec)
			rewind_needed = false;
		else
			rewind_needed = true;
	}

	if (!rewind_needed)
	{
		printf("No rewind required.\n");
		exit(0);
	}
	findLastCheckpoint(datadir_target, divergerec, lastcommontli,
					   &chkptrec, &chkpttli, &chkptredo);
	printf("Rewinding from Last common checkpoint at %X/%X on timeline %u\n",
		   (uint32) (chkptrec >> 32), (uint32) chkptrec,
		   chkpttli);

	/*
	 * Install necessary functions of pg_rewind_support before fetching
	 * the remote file list if a running source server is used for the
	 * rewind operation.
	 */
	if (connstr_source)
		libpqInitSupport();

	/*
	 * Build the filemap, by comparing the remote and local data directories
	 */
	(void) filemap_create();
	fetchRemoteFileList();
	traverse_datadir(datadir_target, &process_local_file);

	/*
	 * Read the target WAL from last checkpoint before the point of fork, to
	 * extract all the pages that were modified on the target cluster after
	 * the fork. We can stop reading after reaching the final shutdown record.
	 * XXX: If we supported rewinding a server that was not shut down cleanly,
	 * we would need to replay until the end of WAL here.
	 */
	extractPageMap(datadir_target, chkptrec, lastcommontli,
				   ControlFile_target.checkPoint);
	filemap_finalize();

	/* XXX: this is probably too verbose even in verbose mode */
	if (verbose)
		print_filemap();

	/* Ok, we're ready to start copying things over. */
	executeFileMap();

	createBackupLabel(chkptredo, chkpttli, chkptrec);

	/*
	 * Update control file of target file and make it ready to
	 * perform archive recovery when restarting.
	 */
	memcpy(&ControlFile, &ControlFile_source, sizeof(ControlFileData));
	ControlFile.minRecoveryPoint = divergerec;
	ControlFile.minRecoveryPointTLI = ControlFile_target.checkPointCopy.ThisTimeLineID;
	ControlFile.state = DB_IN_ARCHIVE_RECOVERY;
	updateControlFile(&ControlFile, datadir_target);

	printf("Done!\n");

	/*
	 * At this point we are basically completely done, so remove the support
	 * bundle from source server if needed.
	 */
	if (connstr_source)
		libpqFinishSupport();
	return 0;
}

static void
sanityChecks(void)
{
	/* Check that there's no backup_label in either cluster */
	/* Check system_id match */
	if (ControlFile_target.system_identifier != ControlFile_source.system_identifier)
	{
		fprintf(stderr, "source and target clusters are from different systems\n");
		exit(1);
	}
	/* check version */
	if (ControlFile_target.pg_control_version != PG_CONTROL_VERSION ||
		ControlFile_source.pg_control_version != PG_CONTROL_VERSION ||
		ControlFile_target.catalog_version_no != CATALOG_VERSION_NO ||
		ControlFile_source.catalog_version_no != CATALOG_VERSION_NO)
	{
		fprintf(stderr, "clusters are not compatible with this version of pg_rewind\n");
		exit(1);
	}

	/*
	 * Target cluster need to use checksums or hint bit wal-logging, this to
	 * prevent from data corruption that could occur because of hint bits.
	 */
	if (ControlFile_target.data_checksum_version != PG_DATA_CHECKSUM_VERSION &&
		!ControlFile_target.wal_log_hints)
	{
		fprintf(stderr, "target master need to use either data checksums or \"wal_log_hints = on\".\n");
		exit(1);
	}

	/*
	 * Target cluster better not be running. This doesn't guard against someone
	 * starting the cluster concurrently. Also, this is probably more strict
	 * than necessary; it's OK if the master was not shut down cleanly, as
	 * long as it isn't running at the moment.
	 */
	if (ControlFile_target.state != DB_SHUTDOWNED &&
		ControlFile_target.state != DB_SHUTDOWNED_IN_RECOVERY)
	{
		fprintf(stderr, "target master must be shut down cleanly.\n");
		exit(1);
	}

    /*
     * When the source is a data directory, also require that the source
     * server is shut down. There isn't any very strong reason for this
     * limitation, but better safe than sorry.
     */
	if (datadir_source &&
		ControlFile_source.state != DB_SHUTDOWNED &&
		ControlFile_source.state != DB_SHUTDOWNED_IN_RECOVERY)
	{
		fprintf(stderr, "source data directory must be shut down cleanly.\n");
		exit(1);
	}
}

/*
 * Determine the TLI of the last common timeline in the histories of the two
 * clusters. *tli is set to the last common timeline, and *recptr is set to
 * the position where the histories diverged (ie. the first WAL record that's
 * not the same in both clusters).
 *
 * Control files of both clusters must be read into ControlFile_target/source
 * before calling this.
 */
static void
findCommonAncestorTimeline(XLogRecPtr *recptr, TimeLineID *tli)
{
	TimeLineID	targettli;
	TimeLineHistoryEntry *sourceHistory;
	int			nentries;
	int			i;
	TimeLineID	sourcetli;

	targettli = ControlFile_target.checkPointCopy.ThisTimeLineID;
	sourcetli = ControlFile_source.checkPointCopy.ThisTimeLineID;

	/* Timeline 1 does not have a history file, so no need to check */
	if (sourcetli == 1)
	{
		sourceHistory = (TimeLineHistoryEntry *) pg_malloc(sizeof(TimeLineHistoryEntry));
		sourceHistory->tli = sourcetli;
		sourceHistory->begin = sourceHistory->end = InvalidXLogRecPtr;
		nentries = 1;
	}
	else
	{
		char		path[MAXPGPATH];
		char	   *histfile;

		TLHistoryFilePath(path, sourcetli);
		histfile = fetchFile(path, NULL);

		sourceHistory = rewind_parseTimeLineHistory(histfile,
													ControlFile_source.checkPointCopy.ThisTimeLineID,
													&nentries);
		pg_free(histfile);
	}

	/*
	 * Trace the history backwards, until we hit the target timeline.
	 *
	 * TODO: This assumes that there are no timeline switches on the target
	 * cluster after the fork.
	 */
	for (i = nentries - 1; i >= 0; i--)
	{
		TimeLineHistoryEntry *entry = &sourceHistory[i];
		if (entry->tli == targettli)
		{
			/* found it */
			*recptr = entry->end;
			*tli = entry->tli;

			free(sourceHistory);
			return;
		}
	}

	fprintf(stderr, "could not find common ancestor of the source and target cluster's timelines\n");
	exit(1);
}


/*
 * Create a backup_label file that forces recovery to begin at the last common
 * checkpoint.
 */
static void
createBackupLabel(XLogRecPtr startpoint, TimeLineID starttli, XLogRecPtr checkpointloc)
{
	XLogSegNo	startsegno;
	char		BackupLabelFilePath[MAXPGPATH];
	FILE	   *fp;
	time_t		stamp_time;
	char		strfbuf[128];
	char		xlogfilename[MAXFNAMELEN];
	struct tm  *tmp;

	if (dry_run)
		return;

	XLByteToSeg(startpoint, startsegno);
	XLogFileName(xlogfilename, starttli, startsegno);

	/*
	 * TODO: move old file out of the way, if any. And perhaps create the
	 * file with temporary name first and rename in place after it's done.
	 */
	snprintf(BackupLabelFilePath, MAXPGPATH,
			 "%s/backup_label" /* BACKUP_LABEL_FILE */, datadir_target);

	/*
	 * Construct backup label file
	 */

	fp = fopen(BackupLabelFilePath, "wb");

	stamp_time = time(NULL);
	tmp = localtime(&stamp_time);
	strftime(strfbuf, sizeof(strfbuf), "%Y-%m-%d %H:%M:%S %Z", tmp);
	fprintf(fp, "START WAL LOCATION: %X/%X (file %s)\n",
			(uint32) (startpoint >> 32), (uint32) startpoint, xlogfilename);
	fprintf(fp, "CHECKPOINT LOCATION: %X/%X\n",
			(uint32) (checkpointloc >> 32), (uint32) checkpointloc);
	fprintf(fp, "BACKUP METHOD: pg_rewind\n");
	fprintf(fp, "BACKUP FROM: standby\n");
	fprintf(fp, "START TIME: %s\n", strfbuf);
	/* omit LABEL: line */

	if (fclose(fp) != 0)
	{
		fprintf(stderr, _("could not write backup label file \"%s\": %s\n"),
				BackupLabelFilePath, strerror(errno));
		exit(2);
	}
}

/*
 * Check CRC of control file
 */
static void
checkControlFile(ControlFileData *ControlFile)
{
	pg_crc32	crc;

	/* Calculate CRC */
	INIT_CRC32(crc);
	COMP_CRC32(crc,
			   (char *) ControlFile,
			   offsetof(ControlFileData, crc));
	FIN_CRC32(crc);

	/* And simply compare it */
	if (!EQ_CRC32(crc, ControlFile->crc))
	{
		fprintf(stderr, "unexpected control file CRC\n");
		exit(1);
	}
}

/*
 * Verify control file contents in the buffer src, and copy it to *ControlFile.
 */
static void
digestControlFile(ControlFileData *ControlFile, char *src, size_t size)
{
	if (size != PG_CONTROL_SIZE)
	{
		fprintf(stderr, "unexpected control file size %d, expected %d\n",
				(int) size, PG_CONTROL_SIZE);
		exit(1);
	}
	memcpy(ControlFile, src, sizeof(ControlFileData));

	/* Additional checks on control file */
	checkControlFile(ControlFile);
}

/*
 * Update a control file with fresh content
 */
static void
updateControlFile(ControlFileData *ControlFile, char *datadir)
{
	char    path[MAXPGPATH];
	char	buffer[PG_CONTROL_SIZE];
	FILE    *fp;

	if (dry_run)
		return;

	/* Recalculate CRC of control file */
	INIT_CRC32(ControlFile->crc);
	COMP_CRC32(ControlFile->crc,
			   (char *) ControlFile,
			   offsetof(ControlFileData, crc));
	FIN_CRC32(ControlFile->crc);

	/*
	 * Write out PG_CONTROL_SIZE bytes into pg_control by zero-padding
	 * the excess over sizeof(ControlFileData) to avoid premature EOF
	 * related errors when reading it.
	 */
	memset(buffer, 0, PG_CONTROL_SIZE);
	memcpy(buffer, ControlFile, sizeof(ControlFileData));

	snprintf(path, MAXPGPATH,
			 "%s/global/pg_control", datadir);

	if ((fp  = fopen(path, "wb")) == NULL)
	{
		fprintf(stderr,"Could not open the pg_control file\n");
		exit(1);
	}

	if (fwrite(buffer, 1,
			   PG_CONTROL_SIZE, fp) != PG_CONTROL_SIZE)
	{
		fprintf(stderr,"Could not write the pg_control file\n");
		exit(1);
	}

	if (fclose(fp))
	{
		fprintf(stderr,"Could not close the pg_control file\n");
		exit(1);
	}
 }
